### ISSUE TITLE
SimplifyBranchSame assumes discriminant is the same as variant index.

### ISSUE THREAD

**Issue by tmiasko** [rust-lang/rust#89485]
```console
$ cat a.rs 
#[derive(Debug, Eq, PartialEq)]
pub enum Type {
    A = 1,
    B = 2,
}
pub fn encode(v: Type) -> Type {
    match v {
        Type::A => Type::B,
        _ => v,
    }
}
fn main() {
  assert_eq!(Type::B, encode(Type::A));
}
$ rustc -Zmir-opt-level=0 a.rs && ./a
$ rustc a.rs && ./a
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `B`,
 right: `A`', a.rs:13:3
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Originally found by @bonega  when experimenting with discriminants starting from 1 in https://github.com/rust-lang/rust/pull/88984#issuecomment-932818833.

@rustbot modify labels: +regression-from-stable-to-stable +I-unsound

**Comment by Urgau** [comment]
After a quick [godbolt](https://godbolt.org/z/Trs1MqKqj) it appears that the regression appear with Rust 1.48.0

**Comment by apiraino** [comment]
A bisection seems to lead to 043f6d747c15068f0053a0542e9b0f17ae7f4de4, nightly 2020-09-26

from e599b53e67ddd197a09a3d8720eed872df481aa0 to 043f6d747c15068f0053a0542e9b0f17ae7f4de4
```
found 11 bors merge commits in the specified range
  commit[0] 2020-09-24UTC: Auto merge of #76918 - ishitatsuyuki:match-fastpath, r=oli-obk
  commit[1] 2020-09-24UTC: Auto merge of #77014 - tmiasko:arena, r=Mark-Simulacrum
  commit[2] 2020-09-25UTC: Auto merge of #77172 - jonas-schievink:rollup-a041rou, r=jonas-schievink
  commit[3] 2020-09-25UTC: Auto merge of #76844 - simonvandel:fix-76803, r=wesleywiser
  commit[4] 2020-09-25UTC: Auto merge of #77144 - flip1995:clippyup, r=Manishearth
  commit[5] 2020-09-25UTC: Auto merge of #77041 - lcnr:const-eval-perf, r=ecstatic-morse
  commit[6] 2020-09-25UTC: Auto merge of #77152 - vandenheuvel:update_chalk_further, r=jackh726
  commit[7] 2020-09-25UTC: Auto merge of #73453 - erikdesjardins:tuplayout, r=eddyb
  commit[8] 2020-09-25UTC: Auto merge of #77157 - tmiasko:simplify-cfg-dup, r=jonas-schievink
  commit[9] 2020-09-25UTC: Auto merge of #77198 - jonas-schievink:rollup-i59i41h, r=jonas-schievink
  commit[10] 2020-09-25UTC: Auto merge of #77201 - matthewjasper:rename-get-unchecked, r=spastorino
```

**Comment by apiraino** [comment]
Assigning priority as discussed in the [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2389485.20SimplifyBranchSame.20assumes.20discriminants.20are.20the.20sam.E2.80.A6) of the Prioritization Working Group.

@rustbot label -I-prioritize +P-critical

**Comment by nbdd0121** [comment]
Looks like it is mostly likely caused by #76844 which is supposed to fix #76803. cc @simonvandel @wesleywiser
